:memo: Improve variables docs (#2070)
platformsh/platformsh-docs: docs/src/development/variables/set-variables.md

---
title: Set variables
description: See how to set variables that you can later use to take control over your app's environment.
---

To set variables, determine which [type of variable](./_index.md#variable-types) to use.
Remember to take into account the order of precedence.
All of the variables can also be [overridden via script](#set-variables-via-script).

## Set variables in your app

Set variables [in code](../../configuration/app/app-reference.md#variables) using the `.platform.app.yaml` file.
These values are the same across all environments and present in the Git repository,
which makes them a poor fit for API keys and other such secrets.

They're better fits for uses such as configuration for consistent builds across every environment,
including setting [PHP configuration values](./_index.md#php-specific-variables).

> Application variables are available at both build time and runtime.

Author: That page is a description of the overall process. I feel like knowing what build time is is a prerequisite for doing things like variables. I'm not sure how much that specific page helps with the understanding of this step in particular. Maybe you could explain what you hope they get out of it?

> Application variables are available at both build time and runtime.

Reviewer: The idea was to help people find the high-level explanation of concepts we bring. Indeed it is a prerequisite, but I thought to point to the right spot in the documentation could help make sure the understanding from the customer perspective is correct. Kind of a reminder?

| --------- | ------------------- | ------- | ----------- |
| JSON      | `--json`            | `false` | Whether the variable is a JSON-serialized value (`true`) or a string (`false`). |
| Sensitive | `--sensitive`       | `false` | If set to `true`, the variable's value is hidden in the console and in CLI responses for added security. It's still readable within the app container. |
| Runtime   | `--visible-runtime` | `true`  | Whether the variable is available at runtime. |
| Build     | `--visible-build`   | `true`  | Whether the variable is available at build time. |

So if you want the `foo` variable to be visible at build time but hidden during runtime, you can set it like this:

```bash
platform variable:create --level project --name foo --value bar --visible-build true --visible-runtime false
```

You can also change the variable options after you create them (except for sensitive values, which can't be set to non-sensitive). For example, to make the `foo` variable visible at runtime and hidden during the build, run this command:

```bash
platform variable:update foo --visible-build false --visible-runtime true
```

> Note that any changes to project variables require you to deploy your environments to have effect.

Reviewer: ```suggestion
Note that any changes to project variables require you to (re)deploy your environments to have effect.
```

### Environment variable options

Environment variables share all of the [options available for project variables](./_index.md#variable-options), with the exception that visibility in the build and runtime can be set only with the CLI (not in the console). Environment variables have one additional option:

| Option      | CLI flag        | Default | Description |
| ----------- | --------------- | ------- | ----------- |
| Inheritable | `--inheritable` | `true`  | Whether the variable is inherited by child environments. |

This option is useful for  setting production-only values such as credentials. For example, to set a PayPal secret value for only the `main` branch and have it not be readable elsewhere, run:

```bash
platform variable:create -e main --name paypal_id --inheritable false --sensitive true
```

Other environments don't inherit it and get either a project variable of the same name if it exists or no value at all.

Note that changing an environment variable causes that environment to be redeployed so the new value is available.
However, child environments are *not* redeployed.

> To make the new value accessible to those environments, redeploy them manually.

Reviewer: ```suggestion
To make the new value accessible to those environments, redeploy them either via:
- [the console](../../administration/web/_index.md),
- [the cli](/cli.md) `platform redeploy -e <name of the child environment>`.
```

Author: We also already have instructions for this: https://docs.platform.sh/development/troubleshoot.html#force-a-redeploy
So I'd just link there.

Reviewer: Edited the suggestion to link to the page you provided ;) thanks!

### Example environment variable

Environment variables are a good place to store values that apply only on Platform.sh and not on your local development environment. This includes API credentials for third-party services, mode settings, and which server (development vs. production) to use.

One example would be to define a Node.js application's build on a production branch (`NODE_ENV=production`),
but use development mode (`NODE_ENV=development`) for each of your development environments.

> Assuming you have a `main` environment for production and then a `staging` environment with more child environments for development,

Reviewer: ```suggestion
Assuming you have a `main` environment for production and a `staging` environment with more child environments for development,
```

value updates trigger a rebuild of the application in the same way that a commit would.

## Set variables via script

You can also provide a `.environment` file as in your application root (next to the `.platform.app.yaml` file for that app).
This file runs as a bash script when the container starts and on all SSH logins.
It can be used to set any environment variables directly, such as the PATH variable.

For example, the following `.environment` file allows any executable installed in the `vendor/bin` directory
(such as executables installed using Composer)
to be run regardless of the current directory:

```bash {location=".environment"}
export PATH=/app/vendor/bin:$PATH
```

Note that the file is sourced after all other environment variables above are defined and so they're available to the script.
This also means the `.environment` script has the last word on environment variable values and can override anything it wants to.

## Map variables

If your app needs different names for environment variable than those set by Platform.sh, which is common for database connections,
map the Platform.sh's variable names to those required by the application.

> Do this in the app with the help of the Config Reader libraries or via a shell script.

Reviewer: ```suggestion
Do this in the app with the help of the [Config Reader libraries](https://github.com/platformsh?q=config-reader), ) or via a shell script.
```

Author: I thought I'd included this :facepalm:

For example, the following [`.environment` script](#set-variables-via-script) exports variables that are visible to the application.
It uses the `jq` library, which is included in all app containers for this purpose.

```bash {location=".environment"}
export APP_DATABASE_HOST=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].host")
export APP_DATABASE_PORT=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].port")
export APP_DATABASE_NAME=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].path")

> export APP_DATABASE_USER=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].username")

Reviewer: ```suggestion
export APP_DATABASE_USERNAME=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].username")
```
same as above

export APP_DATABASE_PASSWORD=$(echo $PLATFORM_RELATIONSHIPS | base64 --decode | jq -r ".database[0].password")
```

> 

Reviewer: ```suggestion
To use it:
1. Add the `.environment` file to your project and adapt the exported variables if needed.
1. [Set the environment variables](#create-environment-specific-variables) (`APP_DATABASE_HOST`, `APP_DATABASE_PORT`, ...).

Now, the values can be accessed in the `env` directly.
```

Reviewer: Or is my understanding of the feature, wrong?

Author: The `.environment` file should set the variables as environment variables on its own. That's what it's there for, as far as I know.
